test/controllers/drivers_controller_test.rb

```diff
@@ -1,175 +1,328 @@
+      expect(new_driver).must_be_nil
+      
+      # Assert re-render
+      must_respond_with :success
     end
   end
   
   describe "edit" do
-    it "responds with success when getting the edit page for an existing, valid driver" do
-      # Arrange
-      # Ensure there is an existing driver saved
+    let (:sample_driver) {
+      Driver.create(name: "Tom Jones", vin: "whatsnewpussycat")
+    }
 
+    it "responds with success when getting the edit page for an existing, valid driver" do
       # Act
-
+      get edit_driver_path(sample_driver.id)
+      
       # Assert
-
+      must_respond_with :success
     end
 
-    it "responds with redirect when getting the edit page for a non-existing driver" do
-      # Arrange
-      # Ensure there is an invalid id that points to no driver
-
+    it "responds with 404 error when getting the edit page for a non-existing driver" do
       # Act
-
+      get edit_driver_path(-1)
+      
       # Assert
-
+      must_respond_with :not_found
     end
   end
 
   describe "update" do
-    it "can update an existing driver with valid information accurately, and redirect" do
+    before do 
+      Driver.create(name: "Tom Jones", vin: "whatsnewpussycat")
+    end 
+
+    let(:driver_hash){
+      {
+        driver: {
+          name: "Tommy Jones",
+          vin: "woahwoahwoahwoah"
+        },
+      }
+    }
+
+    it "can update an existing driver with valid information accurately, and redirects" do
       # Arrange
-      # Ensure there is an existing driver saved
-      # Assign the existing driver's id to a local variable
-      # Set up the form data
+      existing_driver = Driver.first
 
       # Act-Assert
-      # Ensure that there is no change in Driver.count
+      expect {
+        patch driver_path(existing_driver.id), params: driver_hash
+      }.wont_change "Driver.count"
+      
+      # app should take the user back to the driver's show page after the driver is updated
+      must_respond_with :redirect
+      must_redirect_to driver_path(existing_driver.id)
+
+      # update local variable to reflect values in DB
+      existing_driver.reload 
+
+      # values do change
+      expect(existing_driver.name).must_equal driver_hash[:driver][:name]
+      expect(existing_driver.vin).must_equal driver_hash[:driver][:vin]
+    end
 
+    it "does not update any driver if given an invalid id, and responds with a 404" do
+      # Act
+      expect {
+        patch driver_path(-1), params: driver_hash
+      }.wont_change "Driver.count"
+  
       # Assert
-      # Use the local variable of an existing driver's id to find the driver again, and check that its attributes are updated
-      # Check that the controller redirected the user
-
+      must_respond_with :not_found
     end
 
-    it "does not update any driver if given an invalid id, and responds with a 404" do
+    it "does not update driver if the form data is missing a name, and re-renders page successfuly" do
```

Comment:
You two were so awesome and thorough... almost too thorough in the wrong places this time! In your controller tests, you have test cases like "should do this if name is missing, should do this if vin is missing."

Actually, that's very detailed-- checking each attribute for its validity. That's why it's probably more appropriate to put this level of detail in the model test itself. That way, the model tests are responsible for having very detailed tests to prove what's valid or invalid, attribute by attribute.

In contrast, your controller's responsibility is to just have behavior on if your driver was valid or not. (Check your DriversController code -- it doesn't have an if/else or different behavior whether or not the name or vin is missing, it just has behavior on if the Driver is valid or not).

It might be interesting/cleaner to let your controller tests just have "if Driver was valid" and "if Driver was invalid" test cases, and your driver models to check the validity in detail.